### ISSUE TITLE
[macos] Flutter is crash when system upgrade to Big Sur 

### ISSUE THREAD

**Issue by kt71020** [flutter/flutter#70596]
I upgrade system OS to Big Sur, Android studio can't run flutter. 
Using flutter doctor -v  to check problem. "**Bad state: Future already completed**" is present.

```

# flutter doctor -v
[✓] Flutter (Channel stable, 1.22.4, on macOS 11.0.1 20B29 darwin-x64, locale zh-Hant-TW)
    • Flutter version 1.22.4 at /Users/kt/developer/flutter
    • Framework revision 1aafb3a8b9 (3 days ago), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4

⣻Unhandled exception:
Bad state: Future already completed
#0      _AsyncCompleter.complete (dart:async/future_impl.dart:43:31)
#1      _NativeSocket.startConnect.<anonymous closure>.connectNext.<anonymous closure> (dart:io-patch/socket_patch.dart:682:23)
#2      _NativeSocket.issueWriteEvent.issue (dart:io-patch/socket_patch.dart:1102:14)
#3      _NativeSocket.issueWriteEvent (dart:io-patch/socket_patch.dart:1109:12)
#4      _NativeSocket.multiplex (dart:io-patch/socket_patch.dart:1130:11)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)

```

**Comment by TahaTesser** [comment]
Hi @kt71020 
Is this the compete from `flutter doctor -v`? Please provide `doctor -v` logs  Can you please provide `flutter run --verbose`  
Thank you

**Comment by kt71020** [comment]
That is complete from flutter doctor -v.

A part log from flutter run --verbose.
Please see attachment for more log information.

```

```

**Comment by kt71020** [comment]
That is complete from flutter doctor -v.

A part log from flutter run --verbose.
Please see attachment for more log information.

```
flutter run --verbose > flutter_run.log
[  +24 ms]
           #0      throwToolExit (package:flutter_tools/src/base/common.dart:14:3)
           #1      RunCommand.validateCommand (package:flutter_tools/src/commands/run.dart:353:7)
           <asynchronous suspension>
           #2      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:933:11)
           <asynchronous suspension>
           #3      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:836:33)
           <asynchronous suspension>
           #4      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart)
           #5      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:29)
           #6      _rootRun (dart:async/zone.dart:1190:13)
           #7      _CustomZone.run (dart:async/zone.dart:1093:19)
           #8      _runZoned (dart:async/zone.dart:1630:10)
           #9      runZoned (dart:async/zone.dart:1550:10)
           #10     AppContext.run (package:flutter_tools/src/base/context.dart:149:18)
           #11     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:825:20)
           #12     CommandRunner.runCommand (package:args/command_runner.dart:197:27)
           #13     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:335:21)
           #14     _rootRunUnary (dart:async/zone.dart:1198:47)
           #15     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
           #16     _FutureListener.handleValue (dart:async/future_impl.dart:143:18)
           #17     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:696:45)
           #18     Future._propagateToListeners (dart:async/future_impl.dart:725:32)
           #19     Future._completeWithValue (dart:async/future_impl.dart:529:5)
           #20     _AsyncAwaitCompleter.complete (dart:async-patch/async_patch.dart:40:15)
           #21     _completeOnAsyncReturn (dart:async-patch/async_patch.dart:311:13)
           #22     FlutterVersion.checkFlutterVersionFreshness (package:flutter_tools/src/version.dart)
           #23     _rootRunUnary (dart:async/zone.dart:1198:47)
           #24     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
           #25     _FutureListener.handleValue (dart:async/future_impl.dart:143:18)
           #26     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:696:45)
           #27     Future._propagateToListeners (dart:async/future_impl.dart:725:32)
           #28     Future._completeWithValue (dart:async/future_impl.dart:529:5)
           #29     Future._asyncCompleteWithValue.<anonymous closure> (dart:async/future_impl.dart:567:7)
           #30     _rootRun (dart:async/zone.dart:1190:13)
           #31     _CustomZone.run (dart:async/zone.dart:1093:19)
           #32     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
           #33     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
           #34     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
           #35     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
           #36     _runPendingImmediateCallback (dart:isolate-patch/isolate_patch.dart:118:13)
           #37     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:169:5)


Unhandled exception:
Bad state: Future already completed
#0      _AsyncCompleter.complete (dart:async/future_impl.dart:43:31)
#1      _NativeSocket.startConnect.<anonymous closure>.connectNext.<anonymous closure> (dart:io-patch/socket_patch.dart:682:23)
#2      _NativeSocket.issueWriteEvent.issue (dart:io-patch/socket_patch.dart:1102:14)
#3      _NativeSocket.issueWriteEvent (dart:io-patch/socket_patch.dart:1109:12)
#4      _NativeSocket.multiplex (dart:io-patch/socket_patch.dart:1130:11)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
[flutter_run.log](https://github.com/flutter/flutter/files/5546693/flutter_run.log)

```

**Comment by KishorAndroid** [comment]
It seems to me as a Anti-Virus problem after upgrading to macOS Big Sur [https://github.com/flutter/flutter/issues/68999#issuecomment-731528395](url)

**Comment by kt71020** [comment]
> It seems to me as an Anti-Virus problem after upgrading to macOS Big Sur. Quitting Kaspersky solved problem for me.



I do not install Anti-Virus, but **Adguard**. I turn off **Adguard** solved problem for me.

**Comment by omidraha** [comment]
I don't have `Adguard` or `Anti-Virus`,  
But I have the same issue for building only  in `debug` mode on `macOS Big Sur Version 11.0 Beta`.
The  `release` mode  building is OK.

`flutter build ios --debug  --verbose`

**Traceback:**
```
[  +14 ms] Encountered error while building for device.
[   +4 ms] 
           #0      throwToolExit
(package:flutter_tools/src/base/common.dart:14:3)
           #1      BuildIOSCommand.runCommand
           (package:flutter_tools/src/commands/build_ios.dart:103:7)
           <asynchronous suspension>
           #2      FlutterCommand.verifyThenRunCommand
           (package:flutter_tools/src/runner/flutter_command.dart:872:18)
           <asynchronous suspension>
           #3      FlutterCommand.run.<anonymous closure>
           (package:flutter_tools/src/runner/flutter_command.dart:750:33)
           <asynchronous suspension>
           #4      FlutterCommand.run.<anonymous closure>
           (package:flutter_tools/src/runner/flutter_command.dart)
           #5      AppContext.run.<anonymous closure>
           (package:flutter_tools/src/base/context.dart:150:29)
           #6      _rootRun (dart:async/zone.dart:1190:13)
           #7      _CustomZone.run (dart:async/zone.dart:1093:19)
           #8      _runZoned (dart:async/zone.dart:1630:10)
           #9      runZoned (dart:async/zone.dart:1550:10)
           #10     AppContext.run
(package:flutter_tools/src/base/context.dart:149:18)
           #11     FlutterCommand.run
           (package:flutter_tools/src/runner/flutter_command.dart:739:20)
           #12     CommandRunner.runCommand
(package:args/command_runner.dart:197:27)
           #13     FlutterCommandRunner.runCommand.<anonymous closure>
           (package:flutter_tools/src/runner/flutter_command_runner.dart:331:21)
           #14     _rootRunUnary (dart:async/zone.dart:1198:47)
           #15     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
           #16     _FutureListener.handleValue
(dart:async/future_impl.dart:143:18)
           #17     Future._propagateToListeners.handleValueCallback
           (dart:async/future_impl.dart:696:45)
           #18     Future._propagateToListeners
(dart:async/future_impl.dart:725:32)
           #19     Future._completeWithValue (dart:async/future_impl.dart:529:5)
           #20     _AsyncAwaitCompleter.complete
(dart:async-patch/async_patch.dart:40:15)
           #21     _completeOnAsyncReturn
(dart:async-patch/async_patch.dart:311:13)
           #22     FlutterVersion.checkFlutterVersionFreshness
           (package:flutter_tools/src/version.dart)
           #23     _rootRunUnary (dart:async/zone.dart:1198:47)
           #24     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
           #25     _FutureListener.handleValue
(dart:async/future_impl.dart:143:18)
           #26     Future._propagateToListeners.handleValueCallback
           (dart:async/future_impl.dart:696:45)
           #27     Future._propagateToListeners
(dart:async/future_impl.dart:725:32)
           #28     Future._completeWithValue (dart:async/future_impl.dart:529:5)
           #29     Future._asyncCompleteWithValue.<anonymous closure>
           (dart:async/future_impl.dart:567:7)
           #30     _rootRun (dart:async/zone.dart:1190:13)
           #31     _CustomZone.run (dart:async/zone.dart:1093:19)
           #32     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
           #33     _CustomZone.bindCallbackGuarded.<anonymous closure>
           (dart:async/zone.dart:1037:23)
           #34     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
           #35     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
           #36     _runPendingImmediateCallback
           (dart:isolate-patch/isolate_patch.dart:118:13)
           #37     _RawReceivePortImpl._handleMessage
           (dart:isolate-patch/isolate_patch.dart:169:5)
           
           
[   +4 ms] Running shutdown hooks
[        ] Shutdown hook priority 4
[   +1 ms] Shutdown hooks complete
[        ] exiting with code 1
``` 

**Info:**
```
flutter doctor -v
[✓] Flutter (Channel master, 1.22.0, on macOS 11.0 20A5364e, locale en)
    • Flutter version 1.22.0 at /Users/***/ws/flutter
    • Framework revision e64fe11542 (4 months ago), 2020-07-23 18:11:17 -0700
    • Engine revision 626244a72c
    • Dart version 2.9.0 (build 2.9.0-21.0.dev a3815b6590)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/***/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.1

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[!] Proxy Configuration
    • HTTP_PROXY is set
    ! NO_PROXY is not set

[!] Connected device
    ! No devices available

! Doctor found issues in 3 categories.
```

**Comment by kt71020** [comment]
> Flutter plugin not installed; this adds Flutter specific functionality
[see this https://github.com/flutter/flutter/issues/67986](https://github.com/flutter/flutter/issues/67986)

I solve this problem like this
MacOS
```
$ ln -s ~/Library/Application\ Support/Google/AndroidStudio4.1/plugins ~/Library/Application\ Support/AndroidStudio4.1
```
This method works in stable and is a quick fix.

**Comment by omidraha** [comment]
@kt71020,  Flutter and dart plugin issues fixed with your trick, 

But I still have the main issue.
```
flutter doctor -v        
[✓] Flutter (Channel master, 1.22.0, on macOS 11.0 20A5364e, locale en)
    • Flutter version 1.22.0 at /Users/***/***/flutter
    • Framework revision e64fe11542 (4 months ago), 2020-07-23 18:11:17 -0700
    • Engine revision 626244a72c
    • Dart version 2.9.0 (build 2.9.0-21.0.dev a3815b6590)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/***/Library/Android/sdk
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.1

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 50.0.2
    • Dart plugin version 201.9002
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.
```

**Comment by kt71020** [comment]
> Connected device
>     ! No devices available
Google search is great tool !

[https://stackoverflow.com/questions/49045393/flutter-run-no-connected-devices](https://stackoverflow.com/questions/49045393/flutter-run-no-connected-devices)

**Comment by omidraha** [comment]
I think we don't need connected device  when we run this command:

```flutter build ios --debug --verbose```

Because when I build in release mode, there is no connected device too, and It works.:

```flutter build ios --release  ```

**Comment by omidraha** [comment]
For me fixed by running:

```flutter clean```

Before:

```flutter build ios --debug```

**Comment by andreag85** [comment]
None of the above solutions worked for me.
I have Kaspersky Internet Security but disabling the protection don't solve.

**Comment by KishorAndroid** [comment]
> None of the above solutions worked for me.
> I have Kaspersky Internet Security but disabling the protection don't solve.

Disabling won't work. You have to quit Kaspersky application.

**Comment by andersoncarlosfs** [comment]
Hello, 

For those who have installed Avast. Please, you just need to deactivate the option "Scan secure connection" on Web Shield (Core Shields)

Cheers,

**Comment by jonas-zebari** [comment]
Disabling/quitting my antivirus (Sophos) is not an option for me. After upgrading to Big Sur I can no longer develop for iOS. Should this really be closed? Why should people be quitting their antivirus software just to develop in Flutter? I know Big Sur obviously caused this issue, but there has got to be a solution...

**Comment by OliverArthur** [comment]
Hi guys,

I am experiencing the same problem when I try to run the application in vscode and it doesn't work even when I turn off the analytics option `flutter config --no-analytics` 

I still getting this error.
```sh
Launching lib/main.dart on Android SDK built for x86 in debug mode...
lib/main.dart:1
✓ Built build/app/outputs/flutter-apk/app-debug.apk.
Unhandled exception:
Bad state: Future already completed
#0      _AsyncCompleter.complete (dart:async/future_impl.dart:43:31)
#1      _NativeSocket.startConnect.<anonymous closure>.connectNext.<anonymous closure> (dart:io-patch/socket_patch.dart:682:23)
#2      _NativeSocket.issueWriteEvent.issue (dart:io-patch/socket_patch.dart:1102:14)
#3      _NativeSocket.issueWriteEvent (dart:io-patch/socket_patch.dart:1109:12)
#4      _NativeSocket.multiplex (dart:io-patch/socket_patch.dart:1130:11)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)

Exited (255)
```

**Comment by tadadat** [comment]
I don't have antiVirus or AdGuard., I literally just opened this device from shipment, but getting error 
Unable to run 'flutter upgrade' when installing flutter. any idea why?

**Comment by devnullpointer** [comment]
I have this issue on Big Sur when executing flutter run of ios simulator. Turning of Web Shield on Avast allows the execution to work, but its a pain to always turn of web shield when i need to test on simulator.

**Comment by ihorvitruk** [comment]
I faced the same issue, when I migrated to Big Sur. Now I have Big Sur 11.2, but the issue with tests running is still reproducible.

I managed to find temporary workaround for it: to run command with `concurrency` parameter set to 1 (by default it is 10):
``` 
flutter test --concurrency 1
```
In 95% of runs, error does not appear, but in rest 10% - it does. 
I suppose, error is somehow related to **filesystem access** or some **access race condition**. And I do not rely on my temporary solution, because it decrease running time (because only 1 job is used). 

I believe the issue will be investigated and fixed asap.
